File: internal/handlers/link.go
package handlers

import (
	"github.com/gin-gonic/gin"
	"github.com/google/uuid"
	"github.com/rs/zerolog/log"
	"strings"
	"time"
)

func (h *Handlers) cleanupCodes() {
	now := time.Now()
	for k, v := range h.codes {
		if now.After(v.Expires) {
			delete(h.codes, k)
		}
	}
}

func (h *Handlers) GetShortLinkHandler(c *gin.Context) {
	user := h.Hooks.UseUserContext(c)
	if !user.IsLoggedIn {
		c.JSON(401, gin.H{"status": 401, "message": "Unauthorized"})
		return
	}
	h.codeMu.Lock()
	defer h.codeMu.Unlock()
	h.cleanupCodes()
	code := strings.Split(uuid.NewString(), "-")[0]
	h.codes[code] = codeEntry{Issuer: user.Username, Expires: time.Now().Add(time.Hour)}
	c.JSON(200, gin.H{"code": code})
}

type codeRequest struct {
	Code string `json:"code"`
}

type unlinkRequest struct {
	A string `json:"a"`
	B string `json:"b"`
}

func (h *Handlers) UseShortLinkHandler(c *gin.Context) {
	user := h.Hooks.UseUserContext(c)
	if !user.IsLoggedIn {
		c.JSON(401, gin.H{"status": 401, "message": "Unauthorized"})
		return
	}
	var req codeRequest
	if err := c.BindJSON(&req); err != nil {
		c.JSON(400, gin.H{"status": 400, "message": "Bad Request"})
		return
	}
	h.codeMu.Lock()
	h.cleanupCodes()
	entry, ok := h.codes[req.Code]
	if !ok {
		h.codeMu.Unlock()
		c.JSON(400, gin.H{"status": 400, "message": "Invalid code"})
		return
	}
	delete(h.codes, req.Code)
	h.codeMu.Unlock()
	if err := h.LinkDB.AddLink(entry.Issuer, user.Username); err != nil {
		log.Error().Err(err).Msg("failed to add link")
		c.JSON(500, gin.H{"status": 500, "message": "Internal Server Error"})
		return
	}
	c.JSON(200, gin.H{"status": 200, "message": "Linked"})
}

func (h *Handlers) GetLinkedAccountsHandler(c *gin.Context) {
	user := h.Hooks.UseUserContext(c)
	if !user.IsLoggedIn {
		c.JSON(401, gin.H{"status": 401, "message": "Unauthorized"})
		return
	}
	links, err := h.LinkDB.Get(user.Username)
	if err != nil {
		log.Error().Err(err).Msg("failed to get links")
		c.JSON(500, gin.H{"status": 500, "message": "Internal Server Error"})
		return
	}
	c.JSON(200, gin.H{"accounts": links})
}

func (h *Handlers) AdminShowLinkedAccountsHandler(c *gin.Context) {
	user := h.Hooks.UseUserContext(c)
	if !h.isAdmin(user.Email) {
		c.JSON(403, gin.H{"status": 403, "message": "Forbidden"})
		return
	}
	account := c.Query("account")
	links, err := h.LinkDB.Get(account)
	if err != nil {
		log.Error().Err(err).Msg("failed to get links")
		c.JSON(500, gin.H{"status": 500, "message": "Internal Server Error"})
		return
	}
	c.JSON(200, gin.H{"accounts": links})
}

func (h *Handlers) AdminUnlinkAccountsHandler(c *gin.Context) {
	user := h.Hooks.UseUserContext(c)
	if !h.isAdmin(user.Email) {
		c.JSON(403, gin.H{"status": 403, "message": "Forbidden"})
		return
	}
	var req unlinkRequest
	if err := c.BindJSON(&req); err != nil {
		c.JSON(400, gin.H{"status": 400, "message": "Bad Request"})
		return
	}
	if err := h.LinkDB.RemoveLink(req.A, req.B); err != nil {
		log.Error().Err(err).Msg("failed to remove link")
		c.JSON(500, gin.H{"status": 500, "message": "Internal Server Error"})
		return
	}
	c.JSON(200, gin.H{"status": 200, "message": "Unlinked"})
}

func (h *Handlers) isAdmin(email string) bool {
	_, ok := h.admin[strings.ToLower(email)]
	return ok
}

File: internal/handlers/Link.cs
using System;
using System.Collections.Generic;
using Tinylinks.Auth;

namespace Tinylinks.Handlers
{
    public partial class Handlers
    {
        private void CleanupCodes()
        {
            var now = DateTime.UtcNow;
            var remove = new List<string>();
            foreach (var pair in codes)
            {
                if (now > pair.Value.Expires)
                {
                    remove.Add(pair.Key);
                }
            }
            foreach (var key in remove)
            {
                codes.Remove(key);
            }
        }

        public void GetShortLinkHandler(Auth.Context ctx)
        {
            var user = Hooks.UseUserContext(ctx);
            if (!user.IsLoggedIn)
            {
                ctx.JSON(401, new { status = 401, message = "Unauthorized" });
                return;
            }
            string code;
            lock (codeMu)
            {
                CleanupCodes();
                code = Guid.NewGuid().ToString().Split('-')[0];
                codes[code] = new CodeEntry { Issuer = user.Username, Expires = DateTime.UtcNow.AddHours(1) };
            }
            ctx.JSON(200, new { code = code });
        }

        private class CodeRequest
        {
            public string Code { get; set; }
        }

        private class UnlinkRequest
        {
            public string A { get; set; }
            public string B { get; set; }
        }

        public void UseShortLinkHandler(Auth.Context ctx)
        {
            var user = Hooks.UseUserContext(ctx);
            if (!user.IsLoggedIn)
            {
                ctx.JSON(401, new { status = 401, message = "Unauthorized" });
                return;
            }
            CodeRequest req;
            if (!ctx.BindJSON<CodeRequest>(out req))
            {
                ctx.JSON(400, new { status = 400, message = "Bad Request" });
                return;
            }
            CodeEntry entry;
            lock (codeMu)
            {
                CleanupCodes();
                if (!codes.TryGetValue(req.Code, out entry))
                {
                    ctx.JSON(400, new { status = 400, message = "Invalid code" });
                    return;
                }
                codes.Remove(req.Code);
            }
            try
            {
                LinkDB.AddLink(entry.Issuer, user.Username);
            }
            catch (Exception)
            {
                ctx.JSON(500, new { status = 500, message = "Internal Server Error" });
                return;
            }
            ctx.JSON(200, new { status = 200, message = "Linked" });
        }

        public void GetLinkedAccountsHandler(Auth.Context ctx)
        {
            var user = Hooks.UseUserContext(ctx);
            if (!user.IsLoggedIn)
            {
                ctx.JSON(401, new { status = 401, message = "Unauthorized" });
                return;
            }
            string[] links;
            try
            {
                links = LinkDB.Get(user.Username);
            }
            catch (Exception)
            {
                ctx.JSON(500, new { status = 500, message = "Internal Server Error" });
                return;
            }
            ctx.JSON(200, new { accounts = links });
        }

        public void AdminShowLinkedAccountsHandler(Auth.Context ctx)
        {
            var user = Hooks.UseUserContext(ctx);
            if (!IsAdmin(user.Email))
            {
                ctx.JSON(403, new { status = 403, message = "Forbidden" });
                return;
            }
            var account = ctx.Query("account");
            string[] links;
            try
            {
                links = LinkDB.Get(account);
            }
            catch (Exception)
            {
                ctx.JSON(500, new { status = 500, message = "Internal Server Error" });
                return;
            }
            ctx.JSON(200, new { accounts = links });
        }

        public void AdminUnlinkAccountsHandler(Auth.Context ctx)
        {
            var user = Hooks.UseUserContext(ctx);
            if (!IsAdmin(user.Email))
            {
                ctx.JSON(403, new { status = 403, message = "Forbidden" });
                return;
            }
            UnlinkRequest req;
            if (!ctx.BindJSON<UnlinkRequest>(out req))
            {
                ctx.JSON(400, new { status = 400, message = "Bad Request" });
                return;
            }
            try
            {
                LinkDB.RemoveLink(req.A, req.B);
            }
            catch (Exception)
            {
                ctx.JSON(500, new { status = 500, message = "Internal Server Error" });
                return;
            }
            ctx.JSON(200, new { status = 200, message = "Unlinked" });
        }
    }
}
